NetKet/UnitaryTimeEvolution/time_evolution.hpp

```diff
@@ -0,0 +1,142 @@
+#ifndef NETKET_ODESOLVER_HPP
+#define NETKET_ODESOLVER_HPP
+
+#include <memory>
+
+#include "Graph/graph.hpp"
+#include "Hamiltonian/hamiltonian.hpp"
+#include "Hamiltonian/MatrixWrapper/matrix_wrapper.hpp"
+#include "UnitaryTimeEvolution/TimeStepper/time_stepper.hpp"
+#include "Utils/all_utils.hpp"
+
+namespace netket
+{
+
+class TimeEvolutionDriver
+{
+public:
+    using VectorType = Eigen::VectorXcd;
+    using MatrixType = AbstractMatrixWrapper<Hamiltonian>;
+    using StepperType = ode::AbstractTimeStepper<VectorType>;
+
+    static TimeEvolutionDriver FromJson(const json& pars)
+    {
+        Graph graph(pars);
+        Hamiltonian hamiltonian(graph, pars);
+
+        auto pars_te = pars["TimeEvolution"];
+        auto matrix = ConstructMatrixWrapper(pars_te, hamiltonian);
+        auto stepper = ode::ConstructTimeStepper<VectorType>(pars_te, matrix->GetDimension());
+
+        double t0 = pars_te["StartTime"].get<double>();
+        double tmax = pars_te["EndTime"].get<double>();
+        double dt = pars_te["TimeStep"].get<double>();
+
+        return TimeEvolutionDriver(std::move(matrix), std::move(stepper),
+                                   t0, tmax, dt);
+    }
+
+    TimeEvolutionDriver(std::unique_ptr<MatrixType> matrix,
+                        std::unique_ptr<StepperType> stepper,
+                        double t0, double tmax, double dt)
+        : hmat_(std::move(matrix)),
```

Comment:
The point of using `unique_ptr` for these parameters is to signal that this transfers ownership of stepper and matrix wrapper to the TimeEvolutionDriver, which is then also responsible for deallocating them when it is destructed.

After calling `TimeEvolutionDriver(std::move(matrix), ...)` the ownership of `matrix` is transferred to the driver object and the `matrix` unique pointer is guaranteed to be empty. So, as you note, you should not dereference the outer pointer anymore (this would be undefined behaviour), but e.g. [checking whether it is empty](https://en.cppreference.com/w/cpp/memory/unique_ptr/operator_bool) is safe (and will return `true`).

The alternative would be to just pass a raw pointer, but then it would be less clear which part of the code is responsible for deallocating the objects in the end.